src/domain/core/src/services/push_status_service.rs

```diff
@@ -0,0 +1,81 @@
+// Copyright Kamu Data, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::collections::HashMap;
+
+use internal_error::InternalError;
+use opendatafabric::{DatasetHandle, DatasetRefRemote};
+
+use crate::utils::metadata_chain_comparator::CompareChainsResult;
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#[async_trait::async_trait]
+pub trait PushStatusService: Send + Sync {
+    /// For a given dataset, returns list of push remotes with optional
+    /// synchronisation statuses. When could not fetch remote dataset info
+    /// to check state, status is None.
+    ///
+    /// Returns error when failed to determine head block of local dataset.
+    async fn get_statuses(
+        &self,
+        dataset_handle: &DatasetHandle,
+    ) -> Result<DatasetPushStatuses, InternalError>;
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#[derive(Clone, Debug)]
+pub struct DatasetPushStatuses {
+    pub statuses: HashMap<DatasetRefRemote, Option<CompareChainsResult>>,
+}
+
+impl DatasetPushStatuses {
+    pub fn in_sync(&self) -> Vec<&DatasetRefRemote> {
+        self.statuses
+            .iter()
+            .filter_map(|(ds_ref, mb_status)| match mb_status {
+                Some(CompareChainsResult::Equal) => Some(ds_ref),
+                _ => None,
+            })
+            .collect()
+    }
+
+    pub fn out_of_sync(&self) -> Vec<(&DatasetRefRemote, &CompareChainsResult)> {
```

Comment:
This kind of splitting logic seems very specific to how caller wants to process the results and thus probably belongs in the presentation layer rather than in domain.

I think domain services should have very simple return types that have minimal assumptions on how client will be processing them. Hash map in this case seems like an overkill, especially since all you do is iterate it, not access by key.

Personally I would simplify things down to:
```rust
async fn get_statuses(
     &self,
    dataset_handle: &DatasetHandle,
) -> Result<Vec<Result<CompareResult, StatusCheckError>>, InternalError>;
```